### PR TITLE
search: Add support for unindexed structural search

### LINKED ISSUE #8164: Enable structural search on all sourcegraph.com repositories
This is the tracking issue for enabling structural search on all (i.e., uindexed) repositories.

Hit the :+1: if you're interested in this feature.

### PR DESCRIPTION
**Pull request by camdencheek**
This adds support for searching unindexed revisions with a structural
search.

~~Depends on #17924~~
~~Depends on #17930~~
Fixes #8164 

General flow:
- Frontend sends searcher a repo/revision to search with `indexed ==
  false`
- Searcher retrieves a zip of that revision from gitserver or from its
  cache as normal (no change here)
- Searcher unzips the zip, skipping any files that do not match filters
  - We unzip because ripgrep is used to pre-filter the list of files for
    comby, and it doesn't support searching a zip file.
  - We skip files that don't match filters because ripgrep only supports
    glob patterns, and our filters are regex patterns. Additionally,
    this keeps us from having to extract huge archives when we're only
    interested in a couple of files.
- Comby searches the unzipped directory, pre-filtering with ripgrep for
  the sake of performance
- Clean up the unzipped directory
- Matches are returned as normal

### CODE REVIEW

**Comment by keegancsmith** [comment]
Sorry I am running out of time today so won't be able to do a more in-depth review. However, here is a review of the description :)

> - We unzip because ripgrep is used to pre-filter the list of files for comby, and it doesn't support searching a zip file.

Why the use of ripgrep? Searcher already has a regex engine in it that avoids extracting the archives. Did you notice any substantial difference? Including ripgrep means all our docker environments that use searcher/comby will now need it. From a simplicity point of view there is a good argument to stick to searcher's text engine.

FYI I think newer versions of ripgrep may support looking into zip files.

>  - We skip files that don't match filters because ripgrep only supports glob patterns, and our filters are regex patterns. Additionally, this keeps us from having to extract huge archives when we're only interested in a couple of files.

This is another pointer that our built in text search engine may be more appropriate.

**Comment by camdencheek** [comment]
@keegancsmith Thanks for taking a look! 

> Why the use of ripgrep?

@rvantonder can probably add a little more color here since I took an existing spike of his and fleshed it out, but this is my understanding:

1) We need to translate structural search patterns to regex patterns, and this way, we can delegate that responsibility to Comby. Though, now that I think of it, we need to do that translation for Zoekt anyways, so that code [already exists](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/internal/comby/translate.go#L157), making this a non-point. 
2) For some reason it didn't occur to me that we already have a zip regex searcher, so I can't speak to performance of ripgrep vs our searcher, making performance another non-point. The fact that the current searcher doesn't need to extract the archive might make performance an anti-point. 
3) Doing the search ahead of time requires us to pass a (potentially very large) list of files to comby via CLI args. We just did some work elsewhere to avoid constructing this list, so it feels funny to add it back here, but I don't think it's actually a huge performance concern. 

In summary: I don't actually have any good reasons. I'll let @rvantonder chime in before I rip out ripgrep, but using the existing zip searcher seems reasonable to me. 

> FYI I think newer versions of ripgrep may support looking into zip files.

I looked into this, and it turns out that it supports many streaming archive formats, but not zip files (which seems silly since the CLI arg is called `--search-zip`).

**Comment by rvantonder** [comment]
Short version: you're right @keegancsmith, we'd do well to investigate using the searcher regex code, since we didn't do an actual comparison. I propose we create another fork and use searcher/pass file paths and compare numbers. I don't think we shouldn't rule out adding ripgrep until we have some numbers. 

@camdencheek would you mind blocking off 0.5 or 1 day to see how far you get along the lines of using [`regexSearch`](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/cmd/searcher/search/search_regex.go#L298) for structural search, and we can compare numbers?

Long version:

I thought of ripgrep first for two reasons: 

- (a) my intuition is its faster and we discussed the idea before in the context of #11902. In comment we talked about this: https://github.com/sourcegraph/sourcegraph/issues/11902#issuecomment-664885504.  I don't want to take @keegancsmith's quote "I've been wanting to move searcher to ripgrep anyways" out of context, since we probably meant something different from "just plop the ripgrep binary in there" and something more along the lines of bindings/streaming. That said, I think it is worth investigating this route. `ripgrep` as a binary might make it possible to look at a replacement before writing bindings/streaming. 

- (b) implementation is slightly more cumbersome on our regex code, since I added `-rg` as basically a drop-in flag for comby. That's what I started with experimenting initially and I continued along that train of thought. I don't think this is enough reason to rule out using searcher regex. Call it an oversight :-) 

---

From a size perspective, adding ripgrep to our images shouldn't be a difficult configuration thing, and the alpine binary is fairly small (5.06MB), for comparison the `searcher` binary is 37MB on my laptop. We do introduce an external dependency, but if the numbers work out better, then this tradeoff seems worth it to me.

---

Side note: ripgrep doesn't support random-access zip, only streaming formats, like _g_zip. (We also mentioned this possibility  but discovered we're out of luck after looking into it.)

**Comment by camdencheek** [comment]
Nice -- thanks for the additional info. I'll make another branch for comparison

**Comment by camdencheek** [comment]
On first attempt, using `regexSearch` seems to be working quite well. Going to clean it up and make a new PR with that, then do a little bit of perf testing to compare.

**Comment by rvantonder** [comment]
Great! No need to go overboard with the perf testing. If it looks good with some cursory testing, then let's commit to that 👍

**Comment by camdencheek** [comment]
Closing in favor of #17967

File: dev/gqltest/search_test.go [hunk 1]
```diff
@@ -693,7 +693,7 @@ func TestSearch(t *testing.T) {
 				zeroResult: true,
 				wantAlert: &gqltestutil.SearchAlert{
 					Title:       "Unable To Process Query",
-					Description: "Unbalanced expression: unmatched closing parenthesis )",
+					Description: "Unsupported expression. The combination of parentheses in the query have an unclear meaning. Try using the content: filter to quote patterns that contain parentheses",
```
Comment (rvantonder) [inline dev/gqltest/search_test.go:696]: Think you need to rebase? These changes are in `main` :-D

Comment (rvantonder) [reply dev/gqltest/search_test.go:696]: whoops wrong PR :-D

Comment (camdencheek) [reply dev/gqltest/search_test.go:696]: @rvantonder did you review the right PR?

**Comment by keegancsmith** [comment]
Thanks for investigating! Nice work. Will take a look at the other pr now.
